securedrop_client/gui/main.py

```diff
@@ -155,47 +166,28 @@ def on_source_changed(self):
             self.current_source = source_widget.source
             self.show_conversation_for(self.current_source)
 
-    def add_item_content_or(self, adder, item, default):
-        """
-        Private helper function to add correct message to conversation widgets
-        """
-        if item.is_downloaded is False:
-            adder(default)
-        else:
-            adder(item.content)
-
     def show_conversation_for(self, source):
         """
         Show conversation of messages and replies between a source and
         journalists.
         """
-        conversation = ConversationView(self)
-        conversation.setup(self.controller)
-        conversation.add_message('Source name: {}'.format(
-                                 source.journalist_designation))
-
-        # Display each conversation item in the source collection.
-        for conversation_item in source.collection:
-
-            if conversation_item.filename.endswith('msg.gpg'):
-                self.add_item_content_or(conversation.add_message,
-                                         conversation_item,
-                                         "<Message not yet downloaded>")
-            elif conversation_item.filename.endswith('reply.gpg'):
-                self.add_item_content_or(conversation.add_reply,
-                                         conversation_item,
-                                         "<Reply not yet downloaded>")
-            else:
-                conversation.add_file(source, conversation_item)
-
-        container = QWidget()
-        layout = QVBoxLayout()
-        container.setLayout(layout)
-        source_profile = SourceProfileShortWidget(source, self.controller)
-
-        layout.addWidget(source_profile)
-        layout.addWidget(conversation)
-        self.main_view.update_view(container)
+
+        conversation_container = self.conversations.get(source.uuid, None)
+
+        if conversation_container is None:
+            conversation = ConversationView(source, parent=self)
+            conversation.setup(self.controller)
+
+            conversation_container = QWidget()
+            layout = QVBoxLayout()
+            conversation_container.setLayout(layout)
+
+            source_profile = SourceProfileShortWidget(source, self.controller)
+
+            layout.addWidget(source_profile)
+            layout.addWidget(conversation)
+
+        self.main_view.set_conversation(conversation_container)
```

Comment:
I like what you're doing here with the `conversations` dict on the main view :sparkles:. 

The thing is, one case we need to handle is the following:

1. User logs into client.
2. User immediately clicks on source A in the list. At this point, files and messages should not have downloaded/decrypted, i.e. they show `<Message not yet downloaded>`. 
3. We wait for the UI refresh that occurs every six seconds, as by that time our messages should have downloaded and decrypted.

On master, the UI will refresh and we should be able to read the messages in the conversation view. On this branch however, it looks like we need to wait for an API resync, which takes a couple minutes. Can you confirm you see this same behavior?

Recall that resolving the described above is the reason why the UI flicker was introduced in https://github.com/freedomofpress/securedrop-client/pull/142 (to redraw everything just in case the underlying data has changed). Note that clearly we don't want the flicker, just noting this as we want to preserve the behavior on master for the above case just without the API flicker.

Reply:
Yes, this is basically what happens now because this isn't very polished. But if the caching / dict behavior seems to be a thing we want, I can go ahead with the work that actually makes this behave in a clean way.